kratos/spaces/ublas_space.h

```diff
@@ -723,18 +729,29 @@ class UblasSpace
         IndexPartition(system_size).for_each([&](std::size_t Index){
             bool empty = true;
 
-            const std::size_t col_begin = Arow_indices[Index];
-            const std::size_t col_end = Arow_indices[Index + 1];
+            const std::size_t col_begin = r_Arow_indices[Index];
+            const std::size_t col_end = r_Arow_indices[Index + 1];
 
             for (std::size_t j = col_begin; j < col_end; ++j) {
-                if(std::abs(Avalues[j]) > zero_tolerance) {
+                if(std::abs(r_Avalues[j]) > zero_tolerance) {
                     empty = false;
                     break;
                 }
             }
 
             if(empty) {
-                rA(Index, Index) = scale_factor;
+                const auto& it_Acol_indices_row_begin = r_Acol_indices.begin() + col_begin;
+                const auto& it_Acol_indices_row_end = r_Acol_indices.begin() + col_end;
+                const auto pos = std::lower_bound(it_Acol_indices_row_begin, it_Acol_indices_row_end, Index);
+                if (pos != it_Acol_indices_row_end) { // It was found
+                    r_Avalues[pos - it_Acol_indices_begin] = scale_factor;
+            #ifdef KRATOS_DEBUG
+                } else {
+                    KRATOS_ERROR << "Diagonal term (" << Index << ", " << Index << ") is not defined in the system matrix" << std::endl;
+                }
+            #else
+                }
```

Comment:
Not doing anything in release mode is unacceptable to me. At least throw a warning message.